### PR TITLE
Consolidate extended resources and hugepages in Scheduler

### PR DESCRIPTION
**Pull request by aveshagarwal**
Fixes https://github.com/kubernetes/kubernetes/issues/51732

@bsalamat @derekwaynecarr @sjenning @kubernetes/sig-scheduling-pr-reviews

### CODE REVIEW

**Comment by sjenning** [comment]
fyi @ConnorDoyle

File: pkg/api/helper/helpers.go
Comment (vishh) [inline pkg/api/helper/helpers.go]: I don't like this refactoring.
Extended resources != Integer resources in the long run.
I suspect what you are looking for is `IsIntegerResource` which I'd recommend composing

File: pkg/api/v1/helper/helpers.go
Comment (vishh) [inline pkg/api/v1/helper/helpers.go]: What is a generic resource? Instead have an integer resource list for first class resources and use that.

**Comment by bsalamat** [comment]
@ConnorDoyle [was thinking](https://github.com/kubernetes/kubernetes/issues/51732#issuecomment-326399008) about a more generic resource type that could possibly cover all the existing resource types and let us add new ones without the need to change scheduler's code. That's obviously a larger change and requires a design doc first.

This PR is OK and fixes the duplication of code introduced when adding HugePages in the scheduler. The only issue is that I am not sure if "GenericResource" is a good name.

File: pkg/api/v1/helper/helpers.go
Comment (ConnorDoyle) [reply pkg/api/v1/helper/helpers.go]: +1, integer resource predicate can match all extended resources (for now) and the integral first-class resources including hugepages.

**Comment by derekwaynecarr** [comment]
I suggested this name, I am not married to the name, but I am not sure its worth a lot of investment to bikeshed around it versus remove the code duplication.  Calling extended resources + native resources that require no special handling felt generic to me.

**Comment by aveshagarwal** [comment]
> My only ask is to keep the separation between first class and extended resources explicit. 

Just to clarify which one is true: we do not want to combine HugePage with Extended resources? Or we want to combine Extended resources with HugePages but want to go with some other name than  GenericResources? Or we want to combine Extended resources with HugePages but want to continue with the name Extended Resources?

Because if the first one, that means, we would have to continue with HugePage field like this: https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/schedulercache/node_info.go#L75 and I only see marginal improvements even w.r.t code duplication.

**Comment by bsalamat** [comment]
As far as scheduler is concerned, I'd prefer combining `ExtendedResources` with `HugePages` and call them either `ExtendedResources` or `ScalarResources` or `IntegerResources`. I am not sure what are the implications of this approach on the node side though. Derek and Vish should comment on that.

**Comment by ConnorDoyle** [comment]
I am wary of combining the concept of ExtendedResources with a subset of first-class resources in the code. This is mainly because Extended Resources are a documented concept and contributors who have read those docs will have trouble if they are conflated in the code.

How about this plan of action (forgive is restating ground already covered):

- Retain the existing `IsExtendedResource()` predicate.
- Add an `IsIntegerResource()` predicate, which does `IsExtendedResource() || IsHugePageResource()` for now.
- Test `IsIntegerResource()` in the scheduler predicates in place of `IsExtendedResource()`.
- Update the various uses of the ExtendedResources map in scheduler cache types to IntegerResources.

In the future, more unification could be possible to moving some more of the first-class integral resources under the `IsIntegerResource()` predicate and removing the special-case code paths for those resource names. I suggest leaving this as a follow-up after this PR.

**Comment by bsalamat** [comment]
@ConnorDoyle SGTM

**Comment by aveshagarwal** [comment]
@ConnorDoyle Currently IsIntegerResourceName includes standard integer resources and extended resources as shown here: https://github.com/kubernetes/kubernetes/blob/master/pkg/api/helper/helpers.go#L264

So just using IsIntegerResourceName could cause conflicts in Scheduler for resources (like Pods). So I am suggesting to add IsNonStandardIntegerResources as follows:

```
func IsIntegerResourceName(str string) bool {
        return integerResources.Has(str) || IsNonStandardIntegerResources(api.ResourceName(str))
}

func IsNonStandardIntegerResources(str) bool {
IsExtendedResourceName(api.ResourceName(str)) || IsHugePageResourceName(api.ResourceName(str))
}
```
So that this way  IsNonStandardIntegerResources could be used in Scheduler without causing any unintentional issues.

**Comment by ConnorDoyle** [comment]
OK. It seems weird to call hugepages "nonstandard" once they achieve GA but please don't block on this naming nit.

**Comment by aveshagarwal** [comment]
> OK. It seems weird to call hugepages "nonstandard" once they achieve GA but please don't block on this naming nit.

Are you guys @bsalamat  @derekwaynecarr @vishh ok with this? Or do we still want to go with other names(something like @bsalamat's suggestion of ScalarResources looks good too)? Let me know as I'd prefer to avoid further iterations on this? Or if I dont get any response, I'll go ahead with IsNonStandardIntegerResources.

**Comment by bsalamat** [comment]
@aveshagarwal It is unfortunate that cluster objects (pods, configmaps, etc.) are called "resource" in K8s. If we plan to expand the concept and add other resources, like memory, to it in the future, I would suggest using `ScalarResource`. Otherwise, `NonStandardResource` is fine for lack of a better name.

**Comment by vishh** [comment]
ScalarResources or IntegerResources sounds better than NonStandardResources
personally.

**Comment by ConnorDoyle** [comment]
Reals are scalars too, so that doesn't seem as descriptive as the others.

**Comment by aveshagarwal** [comment]
I have updated this PR to consolidate extended and hugepages resources as scalar resources. PTAL.

@sjenning @derekwaynecarr can you check the bug fix commit to make sure its correct?

**Comment by aveshagarwal** [comment]
/release-note-none 

This is only visible in scheduler, and is to avoid code duplication introduced due to separate handling of hugepages from extended resources.

File: plugin/pkg/scheduler/schedulercache/node_info.go
Comment (bsalamat) [inline plugin/pkg/scheduler/schedulercache/node_info.go]: I would just write ScalarResources to cover other scalar resources we add in the future. Besides, OIRs is a bit vague and are deprecated.

File: plugin/pkg/scheduler/schedulercache/node_info.go
Comment (bsalamat) [inline plugin/pkg/scheduler/schedulercache/node_info.go]: Similarly, I would just say: // Lazily allocate scalar resource map.

File: pkg/api/helper/helpers.go
Comment (bsalamat) [inline pkg/api/helper/helpers.go]: s/Hugepages/HugePage/

File: pkg/api/helper/helpers.go
Comment (bsalamat) [inline pkg/api/helper/helpers.go]: Do we need this function here?

Comment (aveshagarwal) [reply pkg/api/helper/helpers.go]: In my understanding, helper/helpers.go and v1/helper/helpers.go have been kept in sync previously, Isn't that case any more?

**Comment by aveshagarwal** [comment]
@bsalamat updated PR, PTAL.

File: pkg/api/helper/helpers.go
Comment (bsalamat) [reply pkg/api/helper/helpers.go]: They don't look exactly in sync, but if that's the convention, it is fine.

**Comment by aveshagarwal** [comment]
@lavalamp could you approve this PR? It has go lgtm and approval from @bsalamat.

**Review by ConnorDoyle — APPROVED** [review]
/lgtm

**Comment by aveshagarwal** [comment]
@smarterclayton can you approve this?

**Pull request merged**
